### PR TITLE
Add upload task context image APIs

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:21621)](https://staging.teamyapp.com/teams/107/tasks/21621)

### CODE REVIEW

File: core/api/gql/schema.graphql [hunk 1]
```diff
@@ -29,6 +29,8 @@ type Mutation {
 	removeAwaitForTask(taskId: ID!, awaitForTaskId: ID!): Task!
 	startDraggingTask(taskId: ID!, clientId: ID!): ID!
 	stopDraggingTask(taskId: ID!, clientId: ID!): ID!
+	createTaskContextFileUploadSession(taskId: ID!): ID!
+	finishTaskContextFileUploadSession(taskId: ID!, fileUploadSessionId: ID!, attachmentListId: ID): Task!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
	createAttachmentListFileUploadSession(attachmentListId: ID!): ID!
	finishAttachmentListUploadSession(fileUploadSessionId: ID!, attachmentListId: ID): AttachmentList!
```

Comment (naturetime10) [reply core/api/gql/schema.graphql]: AttachmentList should be created when creating the task.

File: core/api/gql/schema.graphql [hunk 2]
```diff
@@ -267,6 +269,26 @@ type Task {
 	availableActions: [TaskAction!]!
 	awaitForTasks: [Task!]!
 	links: [TaskLink!]!
+	contextAttachmentList: AttachmentList
+}
+
```
Comment (naturetime10) [inline core/api/gql/schema.graphql:275]:
```suggestion
```

File: core/api/gql/schema.graphql [hunk 3]
```diff
@@ -267,6 +269,26 @@ type Task {
...
+
+type AttachmentList {
+	ownerType: AttachmentListOwnerType!
+	ownerId: ID!
+	listLabel: String!
+	listId: ID!
+	createdAt: Time!
+	updatedAt: Time
+	images: [Image!]!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
	attachments: [Attachment!]!
```

File: core/api/gql/schema.graphql [hunk 4]
```diff
@@ -267,6 +269,26 @@ type Task {
...
+}
+
+type Image {
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
type Attachment {
```

File: core/api/gql/schema.graphql [hunk 5]
```diff
@@ -267,6 +269,26 @@ type Task {
...
+	id: ID!
+	url: String!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql:289]:
```suggestion
    type: AttachmentType!
	url: String!
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/api/gql/schema.graphql (hunk 1 above)
Comment (duanxianbo) [reply core/api/gql/schema.graphql]: I suppose we will do backfill for existing tasks? or we do not care?

File: core/dao/sqldb/attachment.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+type Attachment struct {
+	metrics            dao.Metrics
+	transactionFactory transaction.Factory
+}
+
+var _ dao.Attachment = (*Attachment)(nil)
+
+func (i *Attachment) FindAttachmentsByAttachmentListIDWithTx(ct context.Context, tx *transaction.Transaction, attachmentListID uint64) ([]entity.Attachment, *errs.Error) {
+	var attachments []entity.Attachment
+
```
Comment (naturetime10) [inline core/dao/sqldb/attachment.go]:
```suggestion
```

File: core/dao/sqldb/attachment.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	return attachments, nil
+}
+
+func (i *Attachment) CreateAttachment(ct context.Context, tx *transaction.Transaction, attachment entity.Attachment) *errs.Error {
+	_, err := tx.SQLTx().ExecContext(
+		ct,
+		`
+		INSERT INTO image (id, type, url, size, attachment_list_id, created_at)
```
Comment (naturetime10) [inline core/dao/sqldb/attachment.go]:
```suggestion
		INSERT INTO image 
		(
		    id, 
		    type, 
		    url, 
		    size, 
		    attachment_list_id, 
		    created_at
		  )
```

File: core/dao/attachment_list.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"context"
+
+	"github.com/teamyapp/cloud/libs/errs"
+	"github.com/teamyapp/cloud/libs/transaction"
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type AttachmentList interface {
+	FindAttachmentListByIDWithTx(ct context.Context, tx *transaction.Transaction, attachmentListID uint64) (entity.AttachmentList, *errs.Error)
```
Comment (naturetime10) [inline core/dao/attachment_list.go]:
```suggestion
	FindAttachmentListByIDWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    attachmentListID uint64,
	) (entity.AttachmentList, *errs.Error)
```

File: core/dao/attachment_list.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	FindAttachmentListsByOwnerIDAndOwnerTypeWithTx(ct context.Context, tx *transaction.Transaction, ownerID uint64, ownerType entity.AttachmentListOwnerType) ([]entity.AttachmentList, *errs.Error)
+	FindAttachmentListsWithTx(ct context.Context, tx *transaction.Transaction, ownerID uint64, ownerType entity.AttachmentListOwnerType, ListLabel string) ([]entity.AttachmentList, *errs.Error)
```
Comment (naturetime10) [inline core/dao/attachment_list.go]:
```suggestion
	FindAttachmentListsWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction,
	    ownerType entity.AttachmentListOwnerType, 
	    ownerID uint64, 
	    listLabel string
	 ) ([]entity.AttachmentList, *errs.Error)
```

File: core/dao/attachment.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"context"
+
+	"github.com/teamyapp/cloud/libs/errs"
+	"github.com/teamyapp/cloud/libs/transaction"
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type Attachment interface {
+	FindAttachmentsByAttachmentListIDWithTx(ct context.Context, tx *transaction.Transaction, attachmentListID uint64) ([]entity.Attachment, *errs.Error)
```
Comment (naturetime10) [inline core/dao/attachment.go]:
```suggestion
	FindAttachmentsByAttachmentListIDWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    attachmentListID uint64,
	) ([]entity.Attachment, *errs.Error)
```

File: core/mutation/create_attachment.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+var _ realtime.Mutation = (*CreateAttachment)(nil)
+
+func (c *CreateAttachment) GetID() uint64 {
+	return c.id
+}
+
+func (c *CreateAttachment) Execute(ct context.Context, tx *transaction.Transaction) *errs.Error {
+	internalErr := c.attachmentDao.CreateAttachment(ct, tx, c.attachment)
+	return internalErr
```
Comment (naturetime10) [inline core/mutation/create_attachment.go]:
```suggestion
	return c.attachmentDao.CreateAttachment(ct, tx, c.attachment)
```

File: core/realtime/collection.go [hunk 12]
```diff
@@ -21,4 +21,6 @@ const (
 	TaskActivityCollectionType         CollectionType = "TaskActivity"
 	SprintParticipantCollectionType    CollectionType = "SprintParticipant"
 	StoryCollectionType                CollectionType = "Story"
+	AttachmentListCollectionType       CollectionType = "AttachmentList"
+	ImageCollectionType                CollectionType = "Image"
```
Comment (naturetime10) [inline core/realtime/collection.go]:
```suggestion
	AttachmentCollectionType                CollectionType = "Attachment"
```

File: core/service/attachment.go [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+	err := a.transactionGroupFactory.WithTransactionGroup(ct, true, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		var internalErr *errs.Error
+		attachmentList, internalErr = a.attachmentListDao.FindAttachmentListByIDWithTx(ct, tx, listID)
+		return internalErr
+	})
+
+	return attachmentList, err
+}
+
+func (a *Attachment) FindAttachmentList(ct context.Context, ownerID uint64, ownerType entity.AttachmentListOwnerType, listLabel string) (*entity.AttachmentList, *errs.Error) {
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
func (a *Attachment) FindAttachmentList(
    ct context.Context,
    ownerType entity.AttachmentListOwnerType,
    ownerID uint64,  
    listLabel string,
) (*entity.AttachmentList, *errs.Error) {
```

File: core/service/attachment.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+	var attachmentList *entity.AttachmentList
+	err := a.transactionGroupFactory.WithTransactionGroup(ct, true, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
```
Comment (naturetime10) [inline core/service/attachment.go:52]: Should we make `ownerType, ownerID, listLabel` the primary instead?

File: core/service/attachment.go [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+func (a *Attachment) FindAttachmentsByAttachmentListID(ct context.Context, attachmentListID uint64) ([]entity.Attachment, *errs.Error) {
+	var attachments []entity.Attachment
+	err := a.transactionGroupFactory.WithTransactionGroup(ct, true, func(tx *cloudTransaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		var internalErr *errs.Error
+		attachments, internalErr = a.attachmentDao.FindAttachmentsByAttachmentListIDWithTx(ct, tx, attachmentListID)
+		return internalErr
+	})
+
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
```

File: core/service/attachment.go [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+				return internalErr
+			}
+
+			attachmentList, internalErr = a.attachmentListDao.FindAttachmentListByIDWithTx(ct, tx, attachmentListID)
+			if internalErr != nil {
+				return internalErr
+			}
+
+			attachmentURL := io.GetFileURL(a.cloudWebAPIExternalBaseURL, uploadSession.FileId)
+
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
```

File: core/service/attachment.go [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+			attachment := entity.Attachment{
+				ID:               uploadSession.FileId,
+				Type:             attachmentType,
+				URL:              attachmentURL,
+				Size:             uploadSession.TotalSizeInBytes,
+				AttachmentListID: attachmentListID,
+				CreatedAt:        now,
+			}
+
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
```

File: core/service/attachment.go [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+	"github.com/teamyapp/teamy-backend/core/realtime"
+	"github.com/teamyapp/teamy-backend/core/transaction"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type Attachment struct {
+	cloudClientRegistry            *client.Registry
+	transactionGroupFactory        transaction.GroupFactory
+	cloudWebAPIExternalBaseURL     string
+	logger                         telemetry.Logger
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
	logger                         telemetry.Logger
	cloudClientRegistry            *client.Registry
	transactionGroupFactory        transaction.GroupFactory
	cloudWebAPIExternalBaseURL     string
```

File: core/service/attachment.go [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+		})
+
+	return attachmentList, err
+}
+
+func NewAttachment(
+	cloudClientRegistry *client.Registry,
+	transactionGroupFactory transaction.GroupFactory,
+	cloudWebAPIExternalBaseURL string,
+	logger telemetry.Logger,
```
Comment (naturetime10) [inline core/service/attachment.go]:
```suggestion
	logger telemetry.Logger,
	cloudClientRegistry *client.Registry,
	transactionGroupFactory transaction.GroupFactory,
	cloudWebAPIExternalBaseURL string,
```

File: core/service/task.go [hunk 20]
```diff
@@ -449,6 +458,27 @@ func (t Task) createTask(ct context.Context, teamID uint64, taskInput createTask
...
 
 			rtTx.AppendMutation(createTaskMutation)
+			attachmentList := entity.AttachmentList{
+				OwnerID:   task.ID,
+				OwnerType: entity.AttachmentListOwnerTypeTask,
+				ListID:    genAttachmentListIDRes.UniqueNumber,
+				ListLabel: "context",
+				CreatedAt: now,
+			}
+
```
Comment (naturetime10) [inline core/service/task.go]:
```suggestion
```

File: migrations/20240512212746_create_attachment_tables.sql [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    created_at TIMESTAMP NOT NULL,
+    updated_at TIMESTAMP
+);
+
+CREATE TABLE attachment (
+    id BIGINT NOT NULL PRIMARY KEY,
+    type VARCHAR(50) NOT NULL,
+    url VARCHAR(255) NOT NULL,
+    size BIGINT NOT NULL,
+    attachment_list_id BIGINT NOT NULL REFERENCES attachment_list (list_id),
```
Comment (naturetime10) [inline migrations/20240512212746_create_attachment_tables.sql]:
```suggestion
    attachment_list_id BIGINT NOT NULL REFERENCES attachment_list(list_id),
```

File: migrations/20240512212746_create_attachment_tables.sql [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    created_at TIMESTAMP NOT NULL,
+    updated_at TIMESTAMP
+);
+
+CREATE TABLE attachment_file_upload_session (
+    attachment_list_id BIGINT NOT NULL REFERENCES attachment_list (list_id),
```
Comment (naturetime10) [inline migrations/20240512212746_create_attachment_tables.sql]:
```suggestion
    attachment_list_id BIGINT NOT NULL REFERENCES attachment_list(list_id),
```

File: core/dao/sqldb/attachment_list.go [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+	)
+
+	if err != nil {
+		return entity.AttachmentList{}, errs.NewError(errs.Unknown, err.Error())
+	}
+
+	return attachmentList, nil
+}
+
+func (a *AttachmentList) FindAttachmentListsByOwnerIDAndOwnerTypeWithTx(ct context.Context, tx *transaction.Transaction, ownerID uint64, ownerType entity.AttachmentListOwnerType) ([]entity.AttachmentList, *errs.Error) {
```
Comment (naturetime10) [inline core/dao/sqldb/attachment_list.go]:
```suggestion
func (a *AttachmentList) FindAttachmentListsByOwnerIDAndOwnerTypeWithTx(
    ct context.Context, 
    tx *transaction.Transaction, 
    ownerType entity.AttachmentListOwnerType,
    ownerID uint64,
) ([]entity.AttachmentList, *errs.Error) {
```

File: core/dao/attachment_list.go [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+	"context"
+
+	"github.com/teamyapp/cloud/libs/errs"
+	"github.com/teamyapp/cloud/libs/transaction"
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type AttachmentList interface {
+	FindAttachmentListByIDWithTx(ct context.Context, tx *transaction.Transaction, attachmentListID uint64) (entity.AttachmentList, *errs.Error)
+	FindAttachmentListsByOwnerIDAndOwnerTypeWithTx(ct context.Context, tx *transaction.Transaction, ownerID uint64, ownerType entity.AttachmentListOwnerType) ([]entity.AttachmentList, *errs.Error)
```
Comment (naturetime10) [inline core/dao/attachment_list.go]:
```suggestion
	FindAttachmentListsByOwnerIDAndOwnerTypeWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    ownerType entity.AttachmentListOwnerType,
	    ownerID uint64,
	) ([]entity.AttachmentList, *errs.Error)
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

File: core/service/attachment.go (hunk 14 above)
Comment (duanxianbo) [reply core/service/attachment.go:52]: discussed offline, we control this at app level

**Pull request merged**